Entity/Components/EnergyComponent.gd

```diff
@@ -23,20 +23,20 @@ func use_energy(card: CardBase) -> void:
 	energy -= card.energy_cost
 
 	if energy < 0:
-		energy = 0
+		push_error("energy cannot be negative")
 
 	on_energy_changed.emit()
 
 
 func on_turn_end() -> void:
```

Comment:
this should probably be a `on_turn_start`
reason for that all the other stuff happens at the start of turns. Also, we might have an effect that reduces energy at the start of turns, so it's more logical to put the logic of energy generation at the start
Another problem with that is that since earlier you wrote `starting_energy = 0` and the energy is generated at the end of turns, it means the player has 0 energy for the first turn
Edit : I see start energy was changed to 3, but this doesn't change the idea of putting everything at the start of the turn and start with 0 energy

TL;DR : make this a `on_turn_start`

Reply:
There's already a place to do this too. CardContainer -> _on_player_start_turn

Reply:
yes but your function is `on_turn_end`
do you plan to call `on_turn_end` inside `_on_player_turn_start` ?